matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -163,296 +155,279 @@ public void setMeasurementErrorSigma(double measurementErrorSigma) {
-        int last = gpxList.size() - 1;
+    private List<ViterbiMatchEntry> createviterbiMatchEntries(List<MatchEntry> matchEntries,
+            EdgeFilter edgeFilter) {
+        final List<ViterbiMatchEntry> viterbiMatchEntries = new ArrayList<ViterbiMatchEntry>();
+        ViterbiMatchEntry lastViterbiMatchEntryAdded = null;
+        MatchEntry prevEntry = null;
+        int last = matchEntries.size() - 1;
         for (int i = 0; i <= last; i++) {
-            GPXEntry gpxEntry = gpxList.get(i);
-            if (i == 0 || i == last || distanceCalc.calcDist(
-                    prevEntry.getLat(), prevEntry.getLon(),
-                    gpxEntry.getLat(), gpxEntry.getLon()) > 2 * measurementErrorSigma) {
-                filtered.add(gpxEntry);
-                prevEntry = gpxEntry;
+            MatchEntry matchEntry = matchEntries.get(i);
+            // ignore those which are within 2 * measurementErrorSigma of the previous (though never
+            // ignore the first/last).
+            if (i == 0 || i == last
+                    || distanceCalc.calcDist(prevEntry.gpxEntry.getLat(),
+                            prevEntry.gpxEntry.getLon(), matchEntry.gpxEntry.getLat(),
+                            matchEntry.gpxEntry.getLon()) > 2 * measurementErrorSigma) {
+                lastViterbiMatchEntryAdded = new ViterbiMatchEntry(matchEntry);
+                viterbiMatchEntries.add(lastViterbiMatchEntryAdded);
+                prevEntry = matchEntry;
             } else {
-                logger.debug("Filter out GPX entry: {}", i + 1);
+                matchEntry.markAsNotUsedForMatching();
             }
         }
-        return filtered;
+        return viterbiMatchEntries;
     }
 
     /**
-     * Find the possible locations of each qpxEntry in the graph.
+     * Create candidates per map match event
      */
-    private List<List<QueryResult>> lookupGPXEntries(List<GPXEntry> gpxList,
-                                                     EdgeFilter edgeFilter) {
-
-        List<List<QueryResult>> gpxEntryLocations = new ArrayList<>();
-        for (GPXEntry gpxEntry : gpxList) {
-            gpxEntryLocations.add(locationIndex.findNClosest(gpxEntry.lat, gpxEntry.lon, edgeFilter,
-                    measurementErrorSigma));
+    private void calculateCandidatesPerEvent(List<ViterbiMatchEntry> viterbiMatchEntries,
+            List<QueryResult> allCandidateLocations, QueryGraph queryGraph) {
+
+        // first, find all of the *real* candidate locations for each event i.e. the nodes/edges
+        // that are nearby to the GPX entry location.
+        final EdgeFilter edgeFilter = new DefaultEdgeFilter(
+                algoOptions.getWeighting().getFlagEncoder());
+        final List<List<QueryResult>> candidateLocationsPerEvent = new ArrayList<List<QueryResult>>();
+        for (ViterbiMatchEntry viterbiMatchEntry : viterbiMatchEntries) {
+            // TODO: shouldn't we find those within e.g. 5 * accuracy? Otherwise we're not
```

Comment:
It probably makes more sense to use 2 * accuracy instead because then the GPS error would be within the candidate search radius in 95% instead of 68% of all cases assuming that the normal distribution models GPS noise correctly (see also [the 68–95–99.7 rule](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule)).

In addition to the (increased) search radius radius, I think it makes sense to only consider the k (e.g. 10) closest candidates around the GPS position. This also gives a more predictable map matching run time. The number of candidates and maybe also the search radius / sigma coefficient should be a user configurable parameter. I think this deserves a new issue.

> wouldn't the emission probability be non-zero for points outside measurementSigmaError

Yes. Since the emission probability is modeled as a normal distribution the emission probability is always greater than zero no matter how far away the candidate is from the GPS position. But for practical reasons we have to stop the search at some radius.